google/cloud/bigtable/admin/wait_for_consistency_test.cc

```diff
@@ -0,0 +1,381 @@
+            Status(StatusCode::kUnavailable, "try again"));
+      });
+
+  auto status = AsyncWaitForConsistency(background.cq(), client, table_name,
+                                        token, TestOptions())
+                    .get();
+  EXPECT_THAT(status,
+              StatusIs(StatusCode::kUnavailable, HasSubstr("try again")));
+}
+
+TEST(AsyncWaitForConsistency, NeverConsistent) {
+  std::string const table_name =
+      bigtable::TableName("test-project", "test-instance", "test-table");
+  std::string const token = "test-token";
+
+  internal::AutomaticallyCreatedBackgroundThreads background;
+  auto mock = std::make_shared<MockConnection>();
+  auto client = BigtableTableAdminClient(mock);
+
+  EXPECT_CALL(*mock, AsyncCheckConsistency)
+      .Times(kLimitedErrorCount + 1)
+      .WillRepeatedly([&](btadmin::CheckConsistencyRequest const& request) {
+        EXPECT_EQ(request.name(), table_name);
+        EXPECT_EQ(request.consistency_token(), token);
+        btadmin::CheckConsistencyResponse response;
+        response.set_consistent(false);
+        return make_ready_future(make_status_or(response));
+      });
+
+  auto status = AsyncWaitForConsistency(background.cq(), client, table_name,
+                                        token, TestOptions())
+                    .get();
+  EXPECT_THAT(status, StatusIs(StatusCode::kDeadlineExceeded,
+                               HasSubstr("Polling loop terminated")));
+}
+
+class AsyncWaitForConsistencyCancelTest : public ::testing::Test {
+ protected:
+  int CancelCount() {
+    std::lock_guard<std::mutex> lk(mu_);
+    int n = cancel_count_;
+    cancel_count_ = 0;
+    return n;
+  }
+
+  future<StatusOr<btadmin::CheckConsistencyResponse>> SimulateRequest() {
+    promise<bool> p([this] {
+      std::lock_guard<std::mutex> lk(mu_);
+      ++cancel_count_;
+    });
+    auto f = p.get_future().then([](future<bool> g) {
+      btadmin::CheckConsistencyResponse r;
+      r.set_consistent(g.get());
+      return make_status_or(r);
+    });
+    std::lock_guard<std::mutex> lk(mu_);
+    requests_.push_back(std::move(p));
+    cv_.notify_one();
+    return f;
+  }
+
+  promise<bool> WaitForRequest() {
+    std::unique_lock<std::mutex> lk(mu_);
+    cv_.wait(lk, [&] { return !requests_.empty(); });
+    auto p = std::move(requests_.front());
+    requests_.pop_front();
+    return p;
+  }
+
+ private:
+  // We will simulate the asynchronous requests by pushing promises into this
+  // queue. The test pulls element from the queue to verify the work.
+  std::mutex mu_;
+  std::condition_variable cv_;
+  std::deque<promise<bool>> requests_;
+  int cancel_count_ = 0;
+};
+
+TEST_F(AsyncWaitForConsistencyCancelTest, CancelAndSuccess) {
+  std::string const table_name =
+      bigtable::TableName("test-project", "test-instance", "test-table");
+  std::string const token = "test-token";
+
+  auto fake = std::make_shared<testing_util::FakeCompletionQueueImpl>();
```

Comment:
Instead of using the fake completion queue, consider using the mock:

https://github.com/googleapis/google-cloud-cpp/blob/5d19d3d16eb458414ff96162f3080b09d40369df/google/cloud/testing_util/mock_completion_queue_impl.h#L28

It is easier to grok what it does (we wrote the fake when it was not possible to mock a `google::cloud::CompletionQueue`).

Reply:
See above comment in the `AsyncSequencer` thread

Reply:
SGTM.